OpenOversight/app/email_client.py

```diff
@@ -1,56 +1,165 @@
+            and current_app.config.get(KEY_MAIL_PORT)
+        )
+
+    def send_email(self, email: Email):
+        app = current_app._get_current_object()
+        msg = Message(
+            email.subject,
+            sender=app.config[KEY_OO_SERVICE_EMAIL],
+            recipients=[email.receiver],
+            reply_to=app.config[KEY_OO_HELP_EMAIL],
+        )
+        msg.body = email.body
+        msg.html = email.html
+
+        thread = Thread(
+            target=SMTPEmailProvider.send_async_email,
+            args=[self.mail, app, msg],
+        )
+        current_app.logger.info("Sent email.")
+        thread.start()
+
+    @staticmethod
+    def send_async_email(mail: Mail, app, msg: Message):
+        with app.app_context():
+            mail.send(msg)
 
-    def __new__(cls, config=None, dev=False, testing=False):
-        if (testing or dev) and cls._instance is None:
-            cls._instance = {}
 
-        if cls._instance is None and config:
-            credentials = service_account.Credentials.from_service_account_file(
-                SERVICE_ACCOUNT_FILE, scopes=cls.SCOPES
+class SimulatedEmailProvider(EmailProvider):
+    """Writes messages sent with this provider to log for dev/test usage."""
+
+    def start(self):
+        if not current_app.debug and not current_app.testing:
+            current_app.logger.warning(
+                "Using simulated email provider in non-development environment. "
+                "Please see CONTRIB.md to set up a email provider."
             )
-            delegated_credentials = credentials.with_subject(
-                config[KEY_OO_SERVICE_EMAIL]
+
+    def is_configured(self):
+        return True
+
+    def send_email(self, email: Email):
+        current_app.logger.info("simulated email:\n%s\n%s", email.subject, email.body)
+
+
+class EmailClient:
+    """
+    EmailClient is a Singleton class used for sending email. It auto-detects
+    the email provider implementation based on whether the required
+    configuration is provided for each implementation.
+    """
+
+    DEFAULT_PROVIDER: EmailProvider = SimulatedEmailProvider()
+    PROVIDER_PRECEDENCE: list[EmailProvider] = [
+        GmailEmailProvider(),
+        SMTPEmailProvider(),
+        DEFAULT_PROVIDER,
+    ]
+
+    _provider: Optional[EmailProvider] = None
+    _instance: Optional[Self] = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._provider = cls.auto_detect()
+            cls._provider.start()
+            current_app.logger.info(
+                f"Using email provider: {cls._provider.__class__.__name__}"
             )
-            cls.service = build("gmail", "v1", credentials=delegated_credentials)
             cls._instance = super(EmailClient, cls).__new__(cls)
         return cls._instance
 
+    @classmethod
+    def auto_detect(cls):
+        """Auto-detect the configured email provider to use for email sending."""
+        if current_app.testing:
```

Comment:
`DEBUG` is on in dev mode, so keeping `DEBUG` would prevent local Gmail testing.